tests/generators/test_generator.py

```diff
@@ -0,0 +1,63 @@
+# -*- coding: utf-8 -*-
+"""Tests for the :mod:`aiida_common_workflows.generators.generator` module."""
+import pytest
+from aiida import orm
+from aiida.plugins import WorkflowFactory
+
+
+def test_inputgen_constructor(generate_input_generator_cls):
+    """Test the constructor of a subclass of ``InputGenerator``."""
+
+    cls = generate_input_generator_cls()
+
+    with pytest.raises(RuntimeError):
+        cls()
+
+    cls(process_class=WorkflowFactory('common_workflows.relax.siesta'))
+
+
+def test_get_builder_immutable_kwargs(generate_input_generator_cls, generate_structure):
+    """
+    Test that calling ``get_builder`` does not mutate the ``kwargs`` when they are nodes.
+    In this case a deepcopy would mute the node and therefore return a node with different uuid.
+    """
+
+    cls = generate_input_generator_cls(inputs_dict={'structure': orm.StructureData})
+    generator = cls(process_class=WorkflowFactory('common_workflows.relax.siesta'))
+
+    structure = generate_structure(symbols=('Si',))
+    kwargs = {'structure': structure}
+    generator.get_builder(**kwargs)
+    # The structure is now stored and so should be the same.
+    assert kwargs['structure'].uuid == structure.uuid
+
+
+def test_get_builder_mutable_kwargs(generate_input_generator_cls):
+    """
+    Test that calling ``get_builder`` does mutate the ``kwargs`` when they are not nodes.
+    In this case we want to deepcopy the ``kwargs`` before entering the ``get_builder``.
+    In this specific test, we test a particulat input (``mutable``) that is modified inside
+    the get_builder of ``generate_input_generator_cls``.
+    Here we test that, even if this input is modified inside. This does not affect the
+    original ``kwargs`` since it was deepcopied.
+    """
+
+    cls = generate_input_generator_cls(inputs_dict={'mutable': dict})
+    generator = cls(process_class=WorkflowFactory('common_workflows.relax.siesta'))
+    kwargs = {'mutable': {'test': 333}}
+    generator.get_builder(**kwargs)
+    assert kwargs['mutable'] == {'test': 333}
```

Comment:
It took me a while to understand why this test was actually testing something. It relies on the `_construct_builder` implementation of the class that is returned by the `generate_input_generator_cls` factory. This is not at all clear and the docstring also doesn't say this. For this purpose, I would really advice to have the implementation of `_construct_builder` locally (either in this test file or better even in the test itself). You don't have to copy the entire class and implement it again, you can just monkeypatch the relevant method. I would suggest doing the following:
```python
def test_get_builder_mutable_kwargs(generate_input_generator_cls, monkeypatch):
    
    def construct_builder(self, **kwargs):
        kwargs['mutable']['test'] = 'whatever'
        return self.process_class.get_builder()

    cls = generate_input_generator_cls(inputs_dict={'mutable': dict})
    monkeypatch.setattr(cls, '_construct_builder, construct_builder)
    # and then do the rest of the test.
```

This is a lot clearer and also less susceptible to accidentally breaking the test if the fixture `generate_input_generator_cls` removes the logic that changes the arbitrary keys in the `kwargs` that is not clear at all why it is there.